cpp/frontend_pyreadline.cc

```diff
@@ -4,73 +4,236 @@
+#endif
+
 namespace py_readline {
 
+static Readline* gReadline = nullptr;
+
+// Assuming readline 4.0+
+#if HAVE_READLINE
+
+#define completion_matches(x, y) \
+    rl_completion_matches((x), ((rl_compentry_func_t *)(y)))
+
+static char* do_complete(const char* text, int state) {
+  if (gReadline->completer_ == nullptr) {
+    return nullptr;
+  }
+  rl_attempted_completion_over = 1;
+  Str *gc_text = StrFromC(text);
+  Str *result = completion::ExecuteReadlineCallback(gReadline->completer_, gc_text, state);
+  if (result == nullptr) {
+    return nullptr;
+  }
+
+  // According to https://web.mit.edu/gnu/doc/html/rlman_2.html#SEC37, readline
+  // will free any memory we return to it.
+  int n = len(result) + 1;
+  auto* ret = static_cast<char*>(malloc(n));
+  DCHECK(ret != nullptr);
+  memcpy(ret, result->data(), n);
+  return ret;
+}
+
+static char** completion_handler(const char* text, int start, int end) {
+  rl_completion_append_character = '\0';
+  rl_completion_suppress_append = 0;
+  gReadline->begidx_ = start;
+  gReadline->endidx_ = end;
+  return completion_matches(text, *do_complete);
+}
+
+static void
+display_matches_hook(char **matches, int num_matches, int max_length) {
+  if (gReadline->display_ == nullptr) {
+    return;
+  }
+  auto* gc_matches = Alloc<List<Str*>>();
+  for (int i = 0; i < num_matches; i++) {
+    gc_matches->append(StrFromC(matches[i]));
+  }
+  comp_ui::ExecutePrintCandidates(gReadline->display_, nullptr, gc_matches, max_length);
+}
+
+#endif
+
+Readline::Readline()
+    : GC_CLASS_FIXED(header_, Readline::field_mask(), sizeof(Readline)),
+      begidx_(),
+      endidx_(),
+      completer_delims_(StrFromC(" \t\n`~!@#$%^&*()-=+[{]}\\|;:'\",<>/?")),
+      completer_(),
+      display_() {
+#if HAVE_READLINE
+  using_history();
+  rl_readline_name = "oil";
+  /* Force rebind of TAB to insert-tab */
+  rl_bind_key('\t', rl_insert);
+  /* Bind both ESC-TAB and ESC-ESC to the completion function */
+  rl_bind_key_in_map('\t', rl_complete, emacs_meta_keymap);
+  rl_bind_key_in_map('\033', rl_complete, emacs_meta_keymap);
+  rl_attempted_completion_function = completion_handler;
+  rl_completion_display_matches_hook = display_matches_hook;
+  rl_catch_signals = 0;
+  rl_catch_sigwinch = 0;
+  rl_initialize();
+#else
+  assert(0);  // not implemented
+#endif
+}
+
 void Readline::parse_and_bind(Str* s) {
+#if HAVE_READLINE
+  // Make a copy -- rl_parse_and_bind() modifies its argument
+  Str* copy = StrFromC(s->data(), len(s));
+  rl_parse_and_bind(copy->data());
+#else
   assert(0);  // not implemented
+#endif
 }
 
 void Readline::add_history(Str* line) {
+#if HAVE_READLINE
+  assert(line != nullptr);
+  ::add_history(line->data());
+#else
   assert(0);  // not implemented
+#endif
 }
 
 void Readline::read_history_file(Str* path) {
+#if HAVE_READLINE
+  char* p = nullptr;
+  if (path != nullptr) {
+    p = path->data();
+  }
+  read_history(p);
+#else
   assert(0);  // not implemented
+#endif
```

Comment:
An alternative to this `#if HAVE_READLINE ... #else ...` pattern could be to just keep two implementations in this file, separated by one big if-else. Not sure which is uglier...

Reply:
I'm not too bothered by it ... it would be nice to conditionally compile and link this one file, but we can look at that later